packages/hooks/src/use-boolean.ts

```diff
@@ -22,5 +22,5 @@ export function useBoolean(initialState: InitialState = false) {
     setValue((prev) => !prev)
   }, [])
 
-  return [value, { on, off, toggle }] as const
+  return [value, useRef({ on, off, toggle }).current] as const
```

Comment:
Kindly replace this useRef with a useMemo call, e.g:

```suggestion
  const helpers = useMemo(() => ({ 
    on: ( ) => setValue(true), 
    off: ( ) => setValue(false),
    toggle: () => setValue(prev => !prev)  
  }), [on, off, toggle])
  return [value, helpers] as const
```

Reply:
https://reactjs.org/docs/hooks-reference.html#usememo
> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render

Reply:
Replaced with `useState`

Reply:
Why do we need a guarantee that the helpers object won't ever change their identity?

I think this change is a performance optimization - the hook will work as intended even without any useMemo, useRef or useState usage.

Reply:
To avoid running effects multiple times, and avoid using `exhaustive-deps` from eslint plugin:

```js
const [flag, setFlag] = useBoolean()

useEffect(() => {
  if (foo && flag) {
    // dot sth
  }
}, [foo, flag, setFlag])
```

Reply:
Instead of depending on a stable wrapper object we can extract the needed function and pass them in the dependency array:

```ts
const [flag, { off }] = useBoolean();

React.useEffect(() => {
  if (foo && flag) {
    off();
  }
}, [foo, flag, off]);
```

This way we wouldn't even need to memoize the helper object.
I don't think it is a good idea enforce a guaranteed stable reference for the wrapper object.

---

If your application logic has a dependency on this effect, consider moving this code into an event handler or state machine.

The workaround I currently often use is `useLatestRef` - it's available in the @chakra-ui/hooks package or as [thirdparty package here](https://www.npmjs.com/package/use-latest-ref).

```ts
const [flag, helpers] = useBoolean();
const latestOnFnRef = useLatestRef(helpers.on);

React.useEffect(() => {
  if (foo && flag) {
    latestOnFnRef.current();
  }
}, [foo, flag, latestOnFnRef]);
```